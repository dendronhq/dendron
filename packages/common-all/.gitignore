coverage/
node_modules/
npm-debug.log*
.nyc_output
lib
lib_test
