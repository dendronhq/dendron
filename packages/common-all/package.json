{
  "name": "@dendronhq/common-all",
  "version": "0.124.0",
  "description": "common-all",
  "license": "Apache 2.0",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-all"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com"
  },
  "keywords": [
    "common"
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "echo nop",
    "lint": "echo stub",
    "prepublishOnly": "yarn build",
    "prebuild": "yarn clean && yarn format && yarn lint && echo Using TypeScript && tsc --version",
    "build": "yarn compile",
    "buildCI": "yarn compile",
    "compile": "tsc -p tsconfig.build.json ",
    "watch": "yarn compile --watch"
  },
  "dependencies": {
    "@swc/helpers": "^0.3.2",
    "axios": "^0.21.4",
    "dropbox": "^4.0.30",
    "fast-levenshtein": "^3.0.0",
    "fuse.js": "^6.4.6",
    "github-slugger": "^1.3.0",
    "gray-matter": "^4.0.2",
    "http-status-codes": "^2.1.4",
    "js-yaml": "^4.1.0",
    "lodash": "^4.17.15",
    "luxon": "^1.25.0",
    "minimatch": "5.1.0",
    "nanoid": "^3.1.23",
    "neverthrow": "^5.0.0",
    "normalize-path": "^3.0.0",
    "qs": "^6.10.1",
    "semver": "^7.3.2",
    "spark-md5": "^3.0.2",
    "title": "^3.4.4",
    "vscode-uri": "^3.0.3",
    "zod": "3.19.1",
    "zod-validation-error": "^0.1.1"
  },
  "devDependencies": {
    "@types/fast-levenshtein": "^0.0.2",
    "@types/github-slugger": "^1.3.0",
    "@types/js-yaml": "^4.0.5",
    "@types/lodash": "^4.14.152",
    "@types/luxon": "^1.25.0",
    "@types/minimatch": "^3.0.5",
    "@types/nanoid-dictionary": "^4.2.0",
    "@types/node": "13.11.0",
    "@types/normalize-path": "^3.0.0",
    "@types/semver": "^7.3.4",
    "@types/title": "^3.4.1",
    "@types/unist": "^2.0.3",
    "coveralls": "^3.0.2",
    "rimraf": "^2.6.2",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
