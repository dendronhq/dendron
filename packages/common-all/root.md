---
id: e335a6a8-dca4-4ca5-9842-c6955b953ec1
title: Root
desc: ""
updated: 1609029884026
created: 1609029884026
---
