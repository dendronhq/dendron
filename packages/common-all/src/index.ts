import axios, { AxiosError } from "axios";
import { DateTime } from "luxon";
import minimatch from "minimatch";

export * from "./colors";
export * from "./dnode";
export * from "./helpers";
export * from "./env";
export * from "./assert";
export * from "./uuid";
export * from "./constants";
export * from "./error";
export * from "./time";
export * from "./utils";
export * from "./md";
export * from "./api";
export * from "./vault";
export * from "./VaultUtilsV2";
export * from "./user";
export * from "./analytics";
export * from "./types";
export * from "./themes";
export * from "./FuseEngine";
export * from "./util";
export * from "./timing";
export * from "./config";
export * from "./schema";
export * from "./abTesting";
export * from "./abTests";
export * from "./noteDictsUtils";
export * from "./StatisticsUtils";
export * from "./LabelUtils";
export * from "./store";
export * from "./engine";
export * from "./drivers";
export * from "./DLogger";
export * from "./sidebar";
export * from "./parse";
export * from "./BacklinkUtils";
export * from "./DLinkUtils";
export * from "./service";
export * from "./ResultUtils";
export * as YamlUtils from "./yaml";
export * as GitUtils from "./git";

export { axios, AxiosError };
export { DateTime };

export { minimatch };
