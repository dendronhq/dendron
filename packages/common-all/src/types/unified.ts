// Contains types that help with Dendron's unifiedjs markdown processor.

/** The expected output from the processor, if the processor is used to process or stringify a tree. */
export enum DendronASTDest {
  /**
   * @deprecated - no longer needed since we don't use the markdown preview
   * enhanced anymore
   */
  MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  MD_DENDRON = "MD_DENDRON",
  HTML = "HTML",
}

/**
 * If processor should run in an alternative flavor
 */
export enum ProcFlavor {
  /**
   * No special processing
   */
  REGULAR = "REGULAR",
  /**
   * Apply publishing rules
   */
  PUBLISHING = "PUBLISHING",
  /**
   * Apply preview rules
   */
  PREVIEW = "PREVIEW",
  /**
   * Apply hover preview rules (used for the preview when hovering over a link)
   */
  HOVER_PREVIEW = "HOVER_PREVIEW",

  /**
   * Apply special hover preview rules for the backlinks panel.
   */
  BACKLINKS_PANEL_HOVER = "BACKLINK_HOVER",
}
