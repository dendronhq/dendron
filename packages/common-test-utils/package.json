{
  "name": "@dendronhq/common-test-utils",
  "private": true,
  "version": "0.124.0",
  "description": "",
  "license": "Apache 2.0",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-test-utils"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com"
  },
  "keywords": [],
  "files": [
    "lib"
  ],
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "echo nop",
    "lint": "echo stub",
    "prepublishOnly": "yarn build",
    "prebuild": "yarn clean && yarn format && yarn lint && echo Using TypeScript && tsc --version",
    "build": "yarn compile",
    "buildCI": "yarn compile",
    "compile": "tsc -p tsconfig.json ",
    "coverage": "echo 0 ",
    "watch": "yarn compile --watch"
  },
  "devDependencies": {
    "@types/fs-extra": "^9.0.1",
    "@types/lodash": "^4.14.161",
    "@types/node": "13.11.0",
    "rimraf": "^2.6.2",
    "ts-node": "^8.10.2"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@dendronhq/common-all": "^0.124.0",
    "@dendronhq/common-server": "^0.124.0",
    "@dendronhq/pods-core": "^0.124.0",
    "@types/sinon": "^9.0.9",
    "fs-extra": "^9.0.1",
    "jest": "^28.1.0",
    "lodash": "^4.17.20",
    "sinon": "^9.2.1"
  },
  "gitHead": "d2e0e1c5ada1ac4a507e2da387cf761316de76de"
}
