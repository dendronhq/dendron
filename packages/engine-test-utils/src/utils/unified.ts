import { checkString } from ".";

export class TestUnifiedUtils {
  static verifyPrivateLink = ({
    contents,
    value,
  }: {
    contents: string;
    value: string;
  }) => {
    return checkString(contents, "color: brown", value + " (Private)");
  };
}
