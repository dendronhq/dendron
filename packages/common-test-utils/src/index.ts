import {
  DNodeUtils,
  DVault,
  NoteOpts,
  NoteProps,
  NotePropsByIdDict,
  NoteUtils,
  SchemaModuleOpts,
  SchemaModuleProps,
  SchemaProps,
  SchemaUtils,
  WorkspaceOpts,
  WorkspaceVault,
} from "@dendronhq/common-all";
import {
  note2File,
  resolvePath,
  schemaModuleOpts2File,
  schemaModuleProps2File,
  tmpDir,
  vault2Path,
} from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { PreSetupHookFunctionV4 } from "./types";
import { AssertUtils } from "./utils";
export * from "./fileUtils";
export * from "./noteUtils";
export * from "./presets";
export * from "./types";
export * from "./utils";
export * from "./utilsv2";

export function filterDotFiles(filenames: string[]) {
  return filenames.filter((filename) => !/(^|\/)\.[^\/\.]/g.test(filename));
}

export function getLogFilePath(_name: string) {
  // Placing these in the system temp directory proved difficult, as we both
  // want to generate paths here and pass them from npm in the various LOG_DST
  // environment variables. There's no consistent environment variable we can
  // use for this:
  //
  // * TMPDIR is set for some POSIX-likes, e.g. macOS, but not Linux.
  // * TEMP is set on Windows.
  // @ts-ignore
  //const rootDir = path.dirname(path.dirname(path.dirname(__dirname)));
  //return path.join(rootDir, "logs", `${name}.log`);
  return "stdout";
}

type InitVaultFunc = (vaultPath: string) => void;
export type SetupVaultOpts = {
  vaultDir?: string;
  initDirCb?: (vaultPath: string) => void;
  withAssets?: boolean;
  withGit?: boolean;
};
export type SetupWSOpts = {
  initDirCb?: (vaultPath: string) => void;
  withAssets?: boolean;
  withGit?: boolean;
  wsRoot?: string;
  vaultDir?: string;
};

type SetupVaultsOptsV3 = Omit<SetupVaultOpts, "initDirCb"> & {
  wsRoot: string;
  vaults?: DVault[];
  initVault1?: InitVaultFunc;
  initVault2?: InitVaultFunc;
};

type SetupWSOptsV3 = Omit<SetupVaultsOptsV3, "wsRoot"> & { wsRoot?: string };

/**
 * Relative vaults
 */

export type SetupVaultsOptsV4 = {
  preSetupHook?: PreSetupHookFunctionV4;
  vault: DVault;
};

export class EngineTestUtilsV4 {
  /**
   * Setup a workspace with three vaults
   * The third vault has a different path than name
   */
  static async setupWS(
    opts?: { wsRoot?: string } & {
      setupVaultsOpts?: SetupVaultsOptsV4[];
      singleVault?: boolean;
    }
  ): Promise<WorkspaceOpts> {
    const wsRoot = opts?.wsRoot || tmpDir().name;
    const defaultVaults = opts?.singleVault
      ? ["vault1"]
      : ["vault1", "vault2", "vault3"];
    const setupVaultsOpts: SetupVaultsOptsV4[] =
      opts?.setupVaultsOpts ||
      defaultVaults.map((ent) => ({
        vault: {
          fsPath: ent,
          name: ent === "vault3" ? "vaultThree" : undefined,
        },
        preSetupHook: async ({ vpath, vault, wsRoot }) => {
          const rootModule = SchemaUtils.createRootModule({
            created: 1,
            updated: 1,
            vault,
          });
          await schemaModuleOpts2File(rootModule, vpath, "root");

          const rootNote = await NoteUtils.createRoot({
            created: 1,
            updated: 1,
            vault,
          });
          await note2File({ note: rootNote, vault, wsRoot });
        },
      }));

    const vaults = await Promise.all(
      setupVaultsOpts.flatMap((ent) => {
        return this.setupVault({ ...ent, wsRoot });
      })
    );
    vaults.map((ent) => {
      if (_.isUndefined(ent.name)) {
        delete ent.name;
      }
    });
    return { wsRoot, vaults };
  }

  static async setupVault(opts: SetupVaultsOptsV4 & { wsRoot: string }) {
    const { wsRoot, vault } = opts;
    const vpath = resolvePath(vault.fsPath, wsRoot);
    fs.ensureDirSync(vpath);
    if (opts.preSetupHook) {
      await opts.preSetupHook({ wsRoot, vault, vpath });
    }
    return opts.vault;
  }

  /**
   * Check disk for note
   * @param opts
   * @returns
   */
  static checkVault(
    opts: WorkspaceVault & { match?: string[]; nomatch?: string[] }
  ) {
    const { match, nomatch } = opts;
    const vpath = vault2Path(opts);
    const content = fs.readdirSync(vpath).join("\n");
    return AssertUtils.assertInString({ body: content, match, nomatch });
  }
}

/**
 * Legacy Multi-vault setup
 */
export class EngineTestUtilsV3 {
  static async setupWS(opts: SetupWSOptsV3) {
    const wsRoot = tmpDir().name;
    const vaults = await this.setupVaults({ ...opts, wsRoot });
    return { wsRoot, vaults };
  }

  static async setupVaults(opts: SetupVaultsOptsV3) {
    const { vaults } = _.defaults(opts, {
      vaults: [
        [tmpDir().name, "main"],
        [tmpDir().name, "other"],
      ].map(([vpath, vname]) => {
        return {
          fsPath: path.relative(opts.wsRoot, vpath),
          name: vname,
        };
      }),
    });
    //     {
    //       fsPath: tmpDir().name,
    //       name: "main",
    //     },
    //     {
    //       fsPath: tmpDir().name,
    //       name: "other",
    //     },
    //   ],
    // });
    const cb = [opts.initVault1, opts.initVault2];
    await Promise.all(
      vaults.map(async (ent, idx) => {
        const { fsPath } = ent;
        return EngineTestUtilsV2.setupVault({
          ...opts,
          vaultDir: path.join(opts.wsRoot, fsPath),
          initDirCb: cb[idx],
        });
      })
    );
    return vaults;
  }
}

export class EngineTestUtilsV2 {
  static async setupWS(opts: SetupWSOpts) {
    const { initDirCb, withAssets, withGit } = _.defaults(opts, {
      withAssets: true,
      withGit: true,
    });
    let wsRoot = opts.wsRoot ? opts.wsRoot : tmpDir().name;
    let vaultDir = opts.vaultDir ? opts.vaultDir : path.join(wsRoot, "vault");
    await fs.ensureDir(vaultDir);
    await EngineTestUtilsV2.setupVault({
      vaultDir,
      initDirCb,
      withAssets,
      withGit,
    });
    let vaults = [vaultDir];
    return {
      wsRoot,
      vaults,
    };
  }
  static async setupVault(opts: SetupVaultOpts): Promise<string> {
    const { withAssets, withGit } = opts;
    let vaultDir = opts.vaultDir ? opts.vaultDir : tmpDir().name;
    if (opts?.initDirCb) {
      await opts.initDirCb(vaultDir);
    }
    if (withAssets) {
      const assetsDir = path.join(vaultDir, "assets");
      await fs.ensureDir(assetsDir);
      await fs.ensureFile(path.join(assetsDir, "foo.jpg"));
    }
    if (withGit) {
      fs.ensureDirSync(path.join(vaultDir, ".git"));
    }
    return vaultDir;
  }
}

// === Legacy, deprecate

export class NodeTestUtilsV2 {
  static createNoteProps = async (opts: {
    rootName: string;
    vaultPath: string;
    props?: Partial<NoteProps>;
  }) => {
    const { rootName, vaultPath, props } = opts;
    const vault = { fsPath: vaultPath };
    const foo = NoteUtils.create({
      fname: `${rootName}`,
      id: `${rootName}`,
      created: 1,
      updated: 1,
      children: ["ch1"],
      ...props,
      vault,
    });
    const ch1 = NoteUtils.create({
      fname: `${rootName}.ch1`,
      id: `${rootName}.ch1`,
      created: 1,
      updated: 1,
      vault,
      ...props,
    });
    await note2File({
      note: foo,
      vault: { fsPath: vaultPath },
      wsRoot: "fake_root",
    });
    await note2File({
      note: ch1,
      vault: { fsPath: vaultPath },
      wsRoot: "fake_root",
    });
    return { foo, ch1 };
  };

  static createNote = async (opts: {
    withBody?: boolean;
    vaultDir: string;
    noteProps?: Omit<NoteOpts, "vault"> & { vault?: DVault };
  }): Promise<NoteProps> => {
    const cleanOpts = _.defaults(opts, {
      withBody: true,
      noteProps: [] as NoteOpts[],
    });
    const defaultOpts = {
      created: 1,
      updated: 1,
    };
    const n = cleanOpts.noteProps;
    const body = cleanOpts.withBody ? n.fname + " body" : "";
    const vault = { fsPath: cleanOpts.vaultDir };
    const _n = NoteUtils.create({ ...defaultOpts, body, ...n, vault });
    await note2File({
      note: _n,
      vault: { fsPath: cleanOpts.vaultDir },
      wsRoot: "fake_root",
    });
    return _n;
  };

  static createNotes = async (opts: {
    withBody?: boolean;
    vaultPath: string;
    noteProps?: (Omit<NoteOpts, "vault"> & { vault?: DVault })[];
  }): Promise<NotePropsByIdDict> => {
    const cleanOpts = _.defaults(opts, {
      withBody: true,
      noteProps: [] as NoteOpts[],
    });
    const vault = { fsPath: cleanOpts.vaultPath };
    const defaultOpts = {
      created: 1,
      updated: 1,
    };
    const rootNote = await NoteUtils.createRoot({
      ...defaultOpts,
      vault,
    });
    const out: NotePropsByIdDict = {
      root: rootNote,
    };
    await Promise.all(
      cleanOpts.noteProps.map(async (n) => {
        const body = cleanOpts.withBody ? n.fname + " body" : "";
        const _n = NoteUtils.create({ ...defaultOpts, body, ...n, vault });
        DNodeUtils.addChild(rootNote, _n);
        if (cleanOpts.vaultPath) {
          await note2File({
            note: _n,
            vault: { fsPath: cleanOpts.vaultPath },
            wsRoot: "fake_root",
          });
        }
        out[_n.id] = _n;
        return;
      })
    );
    await note2File({
      note: rootNote,
      vault: { fsPath: cleanOpts.vaultPath },
      wsRoot: "fake_root",
    });
    return out;
  };

  static createSchema = async (opts: {
    vaultDir: string;
    fname: string;
    schemas: SchemaProps[];
  }): Promise<SchemaModuleProps> => {
    const { vaultDir, schemas, fname } = opts;
    const schema = SchemaUtils.createModuleProps({
      fname,
      vault: { fsPath: vaultDir },
    });
    schemas.forEach((s) => {
      schema.schemas[s.id] = s;
    });
    await schemaModuleProps2File(schema, vaultDir, fname);
    return schema;
  };

  static createSchemas = async (opts: {
    vaultPath: string;
    schemaMO?: [SchemaModuleOpts, string][];
  }) => {
    const cleanOpts = _.defaults(opts, {
      schemaMO: [] as [SchemaModuleOpts, string][],
    });
    const { vaultPath, schemaMO } = cleanOpts;
    const vault = { fsPath: vaultPath };
    const rootModule = SchemaUtils.createRootModule({
      created: 1,
      updated: 1,
      vault,
    });
    await schemaModuleOpts2File(rootModule, vaultPath, "root");
    await Promise.all(
      schemaMO.map(async (ent) => {
        const [module, fname] = ent;
        if (vaultPath) {
          await schemaModuleOpts2File(module, vaultPath, fname);
        }
      })
    );
  };

  static createSchemaModuleOpts = async (opts: {
    vaultDir: string;
    rootName: string;
    rootOpts?: Partial<SchemaProps>;
  }) => {
    const { vaultDir, rootName, rootOpts } = opts;
    const vault = { fsPath: vaultDir };
    const schema = SchemaUtils.createFromSchemaOpts({
      fname: `${rootName}`,
      id: `${rootName}`,
      parent: "root",
      created: 1,
      updated: 1,
      children: ["ch1"],
      vault,
      ...rootOpts,
    });
    const ch1 = SchemaUtils.createFromSchemaOpts({
      fname: `${rootName}`,
      vault,
      id: "ch1",
      created: 1,
      updated: 1,
    });
    DNodeUtils.addChild(schema, ch1);
    const schemaModuleProps: [SchemaModuleOpts, string][] = [
      [
        SchemaUtils.createModule({
          version: 1,
          schemas: [schema, ch1],
        }),
        `${rootName}`,
      ],
    ];
    await Promise.all(
      schemaModuleProps.map((ent) => {
        const [module, fname] = ent;
        return schemaModuleOpts2File(module, vaultDir, fname);
      })
    );
    return schemaModuleProps[0][0];
  };

  static normalizeNote({ note }: { note: NoteProps }): Partial<NoteProps> {
    return {
      ..._.omit(note, ["body", "parent", "id", "vault"]),
      body: _.trim(note.body),
    };
  }

  static normalizeNotes(
    notes: NoteProps[] | NotePropsByIdDict
  ): Partial<NoteProps>[] {
    if (!_.isArray(notes)) {
      notes = _.values(notes);
    }
    return notes.map((note) => {
      return NodeTestUtilsV2.normalizeNote({ note });
      //return { ..._.omit(note, ["body", "parent", "id"]), body: _.trim(note.body) };
    });
  }
}
