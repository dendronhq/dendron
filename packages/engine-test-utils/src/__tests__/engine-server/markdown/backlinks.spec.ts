import { ConfigService, ConfigUtils, URI } from "@dendronhq/common-all";
import { AssertUtils, NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { getParsingDependencyDicts, MDUtilsV5 } from "@dendronhq/unified";
import { runEngineTestV5, TestConfigUtils } from "../../..";

describe("backlinks", () => {
  describe("frontmatter tags", () => {
    test("single", async () => {
      await runEngineTestV5(
        async ({ vaults, engine, wsRoot }) => {
          const config = (
            await ConfigService.instance().readConfig(URI.file(wsRoot))
          )._unsafeUnwrap();
          const vault = vaults[0];
          const noteToRender = (
            await engine.findNotes({ fname: "tags.test", vault })
          )[0];
          const noteCacheForRenderDict = await getParsingDependencyDicts(
            noteToRender,
            engine,
            config,
            vaults
          );
          const resp = await MDUtilsV5.procRehypeFull({
            noteToRender,
            noteCacheForRenderDict,
            vault,
            fname: "tags.test",
            config,
            vaults,
          }).process("");
          // should be one backlink
          expect(resp).toMatchSnapshot();
          await AssertUtils.assertInString({
            body: resp.contents as string,
            match: [`<a href="one">One (vault1)</a>`],
          });
        },
        {
          expect,
          preSetupHook: async (opts) => {
            const { wsRoot, vaults } = opts;
            const vault = vaults[0];
            await NoteTestUtilsV4.createNote({
              fname: "one",
              vault,
              wsRoot,
              props: {
                tags: "test",
              },
            });
            await NoteTestUtilsV4.createNote({
              fname: "tags.test",
              vault,
              wsRoot,
            });
            await TestConfigUtils.withConfig(
              (config) => {
                const DefaultConfig = ConfigUtils.genDefaultConfig();
                ConfigUtils.setVaults(
                  DefaultConfig,
                  ConfigUtils.getVaults(config)
                );
                return DefaultConfig;
              },
              { wsRoot }
            );
          },
        }
      );
    });

    test("multiple", async () => {
      await runEngineTestV5(
        async ({ vaults, engine, wsRoot }) => {
          const config = (
            await ConfigService.instance().readConfig(URI.file(wsRoot))
          )._unsafeUnwrap();
          const vault = vaults[0];
          const noteToRender = (
            await engine.findNotes({ fname: "tags.test", vault })
          )[0];
          const noteCacheForRenderDict = await getParsingDependencyDicts(
            noteToRender,
            engine,
            config,
            vaults
          );
          const resp = await MDUtilsV5.procRehypeFull({
            noteToRender,
            noteCacheForRenderDict,
            vault,
            fname: "tags.test",
            config,
            vaults,
          }).process("");
          // should be one backlink
          expect(resp).toMatchSnapshot();
          await AssertUtils.assertInString({
            body: resp.contents as string,
            match: [
              `<a href="one">One (vault1)</a>`,
              `<a href="two">Two (vault1)</a>`,
            ],
          });
        },
        {
          expect,
          preSetupHook: async (opts) => {
            const { wsRoot, vaults } = opts;
            const vault = vaults[0];
            await NoteTestUtilsV4.createNote({
              fname: "one",
              vault,
              wsRoot,
              props: {
                tags: ["test", "other"],
              },
            });

            await NoteTestUtilsV4.createNote({
              fname: "two",
              vault,
              wsRoot,
              props: {
                tags: "test",
              },
            });
            await NoteTestUtilsV4.createNote({
              fname: "tags.test",
              vault,
              wsRoot,
            });
          },
        }
      );
    });
  });
  //
  test("hashtag", async () => {
    await runEngineTestV5(
      async ({ vaults, engine, wsRoot }) => {
        const config = (
          await ConfigService.instance().readConfig(URI.file(wsRoot))
        )._unsafeUnwrap();
        const vault = vaults[0];
        const noteToRender = (
          await engine.findNotes({ fname: "tags.test", vault })
        )[0];
        const noteCacheForRenderDict = await getParsingDependencyDicts(
          noteToRender,
          engine,
          config,
          vaults
        );
        const resp = await MDUtilsV5.procRehypeFull({
          noteToRender,
          noteCacheForRenderDict,
          vault,
          fname: "tags.test",
          config,
          vaults,
        }).process("");
        // should be one backlink
        expect(resp).toMatchSnapshot();
        expect(
          await AssertUtils.assertInString({
            body: resp.contents as string,
            match: [`<a href="one">One (vault1)</a>`],
          })
        ).toBeTruthy();
      },
      {
        expect,
        preSetupHook: async (opts) => {
          const { wsRoot, vaults } = opts;
          const vault = vaults[0];
          await NoteTestUtilsV4.createNote({
            fname: "one",
            vault,
            wsRoot,
            body: "#test",
          });
          await NoteTestUtilsV4.createNote({
            fname: "tags.test",
            vault,
            wsRoot,
          });
        },
      }
    );
  });
});
