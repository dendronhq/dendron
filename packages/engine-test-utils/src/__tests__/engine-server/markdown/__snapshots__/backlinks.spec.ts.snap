// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backlinks frontmatter tags multiple 1`] = `
VFile {
  "contents": "<h1 id=\\"test\\">Test</h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"one.html\\">One (vault1)</a></li>
<li><a href=\\"two.html\\">Two (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`backlinks frontmatter tags single 1`] = `
VFile {
  "contents": "<h1 id=\\"test\\">Test</h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"one.html\\">One (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`backlinks hashtag 1`] = `
VFile {
  "contents": "<h1 id=\\"test\\">Test</h1>
<hr>
<strong>Backlinks</strong>
<ul>
<li><a href=\\"one\\">One (vault1)</a></li>
</ul>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
